src/utils/conversion.rs

```diff
@@ -38,3 +49,451 @@ where
+            };
+            yield Ok(buffer.to_vec().into());
+
+            let mut last_byte = unsafe { *input.next().wrapping_sub(1) };
+            if pos == 0 {
+                /* already at byte boundary, or last file: write last byte */
+                yield Ok(vec![last_byte].into());
+            } else {
+                /* append empty blocks to last byte */
+                last_byte &= (0x100u16 >> pos) as u8 - 1; /* assure unused bits are zero */
+                if pos & 1 > 0 {
+                    if pos == 1 {
+                        yield Ok(vec![last_byte, 0, 0, 0, 0xff, 0xff].into());
+                    } else {
+                        yield Ok(vec![last_byte, 0, 0, 0xff, 0xff].into());
+                    }
+                } else {
+                    /* even -- append 1, 2, or 3 empty fixed blocks */
+                    match pos {
+                        6 => {
+                            yield Ok(vec![last_byte | 8, 0x20, 0x80, 0].into());
+                        }
+                        4 => {
+                            yield Ok(vec![last_byte | 0x20, 0x80, 0].into());
+                        }
+                        2 => {
+                            yield Ok(vec![last_byte | 0x80, 0].into());
+                        }
+                        _ => {
+                            yield Err(anyhow!("Incorrect POS"));
+                            return;
+                        }
+                    }
+                }
+            }
+
+            /* update crc and tot */
+            crc = unsafe { crc32_combine(crc, input.get4().await?, len.try_into()?) };
+            let _ = input.get4().await?;
+            tot += len as u32;
+            /* clean up */
+            unsafe { inflateEnd(&mut strm_wrap.0) };
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rand::RngCore;
+    use tokio::io::AsyncBufReadExt;
+
+    use super::*;
+
+    fn generate_data(data: Vec<Vec<u8>>) -> impl Stream<Item = Vec<u8>> {
+        stream! {
+            for datum in data {
+                let encoder = GzipEncoder::with_quality(&datum[..], async_compression::Level::Fastest);
+                let buf = BufReader::with_capacity(100_000_000, encoder).fill_buf().await.unwrap().to_vec();
+                yield buf;
+            }
+        }
+    }
+
+    async fn pack_join_unpack(testcase: Vec<Vec<u8>>) {
+        let gt = testcase.concat();
+        let stream = join_gzip(generate_data(testcase));
+        pin_mut!(stream);
+        let reader = StreamReader::new(
+            stream.map(|result| std::io::Result::Ok(Bytes::from_owner(result.unwrap()))),
+        );
+        let decoder = GzipDecoder::new(reader);
+        let mut buf_reader = BufReader::with_capacity(100_000_000, decoder);
+        let result = buf_reader.fill_buf().await.unwrap();
+        assert_eq!(result.to_vec(), gt);
+    }
+
+    #[tokio::test]
+    async fn test_smoke() {
+        let input = vec![vec![1u8, 2], vec![3, 4]];
```

Comment:
I would also check how it works with empty chunks, even if we don't encounter them on practice